src/coreclr/vm/jitinterface.cpp

```diff
@@ -8648,15 +8648,11 @@ bool CEEInfo::resolveVirtualMethodHelper(CORINFO_DEVIRTUALIZATION_INFO * info)
             {
                 _ASSERTE(pObjMT->IsArray());
 
-                // We cannot devirtualize unless we know the exact array element type
-                //
-                TypeHandle elemType = pObjMT->GetArrayElementTypeHandle();
-                if (elemType.IsCanonicalSubtype())
-                {
-                    info->detail = CORINFO_DEVIRTUALIZATION_FAILED_LOOKUP;
-                    return false;
-                }
-                pDevirtMD = GetActualImplementationForArrayGenericIListOrIReadOnlyListMethod(pBaseMD, elemType);
+                // The instantiation we want is based on the interface element type, not the
+                // array element type.
+                TypeHandle resultElemType = pBaseMT->GetInstantiation()[0];
+                _ASSERTE(!resultElemType.IsCanonicalSubtype());
```

Comment:
How do we prevent this from being a canonical subtype?  For instance, imagine we're running a function like this... And we call it with an array of Derived, but T is Base.
```
class Base{}
class Derived : Base {}

void Foo<T>(object[] x)
{
    while (i < 100_000)
    {
        i++;
        var values = x.Cast<T>();
        foreach (var value in values)
        {
        }
    }
}
```

Reply:
Ah I thought we ruled it out up above, but we don't. Let me fix that.

Reply:
Reworked things and added this as a second test case.